// @unocss/preset-mini/dist/theme.cjs
'use strict';

const colors = require('./colors.cjs');
const _default = require('./shared/preset-mini.DLRoYX1Y.cjs');
require('./shared/preset-mini.hSTXklYn.cjs');
require('./shared/preset-mini.BCYTcd-i.cjs');
require('@unocss/core');
require('@unocss/rule-utils');



exports.colors = colors.colors;
exports.baseSize = _default.baseSize;
exports.blur = _default.blur;
exports.borderRadius = _default.borderRadius;
exports.boxShadow = _default.boxShadow;
exports.breakpoints = _default.breakpoints;
exports.containers = _default.containers;
exports.dropShadow = _default.dropShadow;
exports.duration = _default.duration;
exports.easing = _default.easing;
exports.fontFamily = _default.fontFamily;
exports.fontSize = _default.fontSize;
exports.fontWeight = _default.fontWeight;
exports.height = _default.height;
exports.letterSpacing = _default.letterSpacing;
exports.lineHeight = _default.lineHeight;
exports.lineWidth = _default.lineWidth;
exports.maxHeight = _default.maxHeight;
exports.maxWidth = _default.maxWidth;
exports.media = _default.media;
exports.preflightBase = _default.preflightBase;
exports.ringWidth = _default.ringWidth;
exports.spacing = _default.spacing;
exports.textIndent = _default.textIndent;
exports.textShadow = _default.textShadow;
exports.textStrokeWidth = _default.textStrokeWidth;
exports.theme = _default.theme;
exports.verticalBreakpoints = _default.verticalBreakpoints;
exports.width = _default.width;
exports.wordSpacing = _default.wordSpacing;
exports.zIndex = _default.zIndex;
