// @unocss/preset-mini/dist/utils.cjs
'use strict';

const utilities = require('./shared/preset-mini.BCYTcd-i.cjs');
const ruleUtils = require('@unocss/rule-utils');
require('@unocss/core');



exports.CONTROL_MINI_NO_NEGATIVE = utilities.CONTROL_MINI_NO_NEGATIVE;
exports.colorResolver = utilities.colorResolver;
exports.colorableShadows = utilities.colorableShadows;
exports.cornerMap = utilities.cornerMap;
exports.cssMathFnRE = utilities.cssMathFnRE;
exports.directionMap = utilities.directionMap;
exports.directionSize = utilities.directionSize;
exports.globalKeywords = utilities.globalKeywords;
exports.h = utilities.h;
exports.handler = utilities.handler;
exports.hasParseableColor = utilities.hasParseableColor;
exports.insetMap = utilities.insetMap;
exports.isCSSMathFn = utilities.isCSSMathFn;
exports.isSize = utilities.isSize;
exports.makeGlobalStaticRules = utilities.makeGlobalStaticRules;
exports.parseColor = utilities.parseColor;
exports.positionMap = utilities.positionMap;
exports.resolveBreakpoints = utilities.resolveBreakpoints;
exports.resolveVerticalBreakpoints = utilities.resolveVerticalBreakpoints;
exports.splitShorthand = utilities.splitShorthand;
exports.transformXYZ = utilities.transformXYZ;
exports.valueHandlers = utilities.valueHandlers;
exports.xyzArray = utilities.xyzArray;
exports.xyzMap = utilities.xyzMap;
Object.prototype.hasOwnProperty.call(ruleUtils, '__proto__') &&
	!Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
	Object.defineProperty(exports, '__proto__', {
		enumerable: true,
		value: ruleUtils['__proto__']
	});

Object.keys(ruleUtils).forEach(function (k) {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = ruleUtils[k];
});
