// unplugin-auto-import/dist/chunk-KSE6FVK6.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }



var _chunk26H2SXXWjs = require('./chunk-26H2SXXW.js');

// src/index.ts
var _fs = require('fs');
var _unplugin = require('unplugin');
var _utils = require('@antfu/utils');

// src/core/options.ts
var _path = require('path');

var _pluginutils = require('@rollup/pluginutils');
var _localpkg = require('local-pkg');
function resolveOptions(options = {}, root = process.cwd()) {
  const imports = flattenImportsMap(options.imports, options.presetOverriding);
  const {
    dts = _localpkg.isPackageExists.call(void 0, "typescript")
  } = options;
  const eslintrc = options.eslintrc || {};
  eslintrc.enabled = eslintrc.enabled === void 0 ? false : eslintrc.enabled;
  eslintrc.filepath = eslintrc.filepath || "./.eslintrc-auto-import.json";
  eslintrc.globalsPropValue = eslintrc.globalsPropValue === void 0 ? true : eslintrc.globalsPropValue;
  const resolved = _chunk26H2SXXWjs.__spreadProps.call(void 0, _chunk26H2SXXWjs.__spreadValues.call(void 0, {
    sourceMap: false,
    resolvedImports: {},
    presetOverriding: false,
    ignore: []
  }, options), {
    dts: dts === false ? false : dts === true ? _path.resolve.call(void 0, root, "auto-imports.d.ts") : _path.resolve.call(void 0, root, dts),
    imports,
    resolvers: options.resolvers ? [options.resolvers].flat(2) : [],
    idFilter: _pluginutils.createFilter.call(void 0, options.include || [/\.[jt]sx?$/, /\.vue$/, /\.vue\?vue/, /\.svelte$/], options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/]),
    eslintrc
  });
  return resolved;
}
function flattenImportsMap(map, overriding = false) {
  const flat = {};
  _utils.toArray.call(void 0, map).forEach((definition) => {
    if (typeof definition === "string") {
      if (!_chunk26H2SXXWjs.presets[definition])
        throw new Error(`[auto-import] preset ${definition} not found`);
      const preset = _chunk26H2SXXWjs.presets[definition];
      definition = typeof preset === "function" ? preset() : preset;
    }
    for (const mod of Object.keys(definition)) {
      for (const id of definition[mod]) {
        const meta = {
          path: mod
        };
        let name;
        if (Array.isArray(id)) {
          name = id[1];
          meta.importName = id[0];
        } else {
          name = id;
          meta.importName = id;
        }
        if (flat[name] && !overriding)
          throw new Error(`[auto-import] identifier ${name} already defined with ${flat[name].path}`);
        flat[name] = meta;
      }
    }
  });
  return flat;
}

// src/core/transform.ts
var _magicstring = require('magic-string'); var _magicstring2 = _interopRequireDefault(_magicstring);
var excludeRE = [
  /\bimport\s*([\s\S]+?)\s*from\b/g,
  /\bfunction\s*([\w_$]+?)\s*\(/g,
  /\b(?:const|let|var)\s+?(\[[\s\S]*?\]|\{[\s\S]*?\}|[\s\S]+?)\s*?[=;\n]/g
];
var matchRE = /(?<![\w_$/)]\.)([\w_$]+?)[^\w_${]/g;
var regexRE = /\/.*?(?<!\\)\/[gimsuy]*/g;
var importAsRE = /^.*\sas\s+/;
var seperatorRE = /[,[\]{}\n]/g;
var multilineCommentsRE = /\/\*\s(.|[\r\n])*?\*\//gm;
var singlelineCommentsRE = /\/\/\s.*/g;
var templateLiteralRE = /\$\{(.*?)\}/g;
var quotesRE = [
  /(["'])((?:\\\1|(?!\1)|.|\r)*?)\1/gm,
  /([`])((?:\\\1|(?!\1)|.|\n|\r)*?)\1/gm
];
function stripeCommentsAndStrings(code) {
  return code.replace(multilineCommentsRE, "").replace(singlelineCommentsRE, "").replace(templateLiteralRE, "` + $1 + `").replace(quotesRE[0], '""').replace(quotesRE[1], "``").replace(regexRE, 'new RegExp("")');
}
async function transform(code, id, {
  imports,
  sourceMap,
  resolvers,
  resolvedImports = {},
  ignore = []
}) {
  const striped = stripeCommentsAndStrings(code);
  const identifiers = new Set(Array.from(striped.matchAll(matchRE)).map((i) => i[1]));
  ignore.forEach((i) => {
    if (typeof i === "string") {
      identifiers.delete(i);
    } else {
      identifiers.forEach((id2) => {
        if (id2.match(i))
          identifiers.delete(id2);
      });
    }
  });
  if (!identifiers.size)
    return null;
  for (const regex of excludeRE) {
    Array.from(striped.matchAll(regex)).flatMap((i) => {
      var _a, _b;
      return [
        ...((_a = i[1]) == null ? void 0 : _a.split(seperatorRE)) || [],
        ...((_b = i[2]) == null ? void 0 : _b.split(seperatorRE)) || []
      ];
    }).map((i) => i.replace(importAsRE, "").trim()).forEach((i) => identifiers.delete(i));
  }
  if (!identifiers.size)
    return null;
  const modules = {};
  const addToModules = (info) => {
    if (!modules[info.path])
      modules[info.path] = [info];
    else
      modules[info.path].push(info);
  };
  for (const name of Array.from(identifiers)) {
    let info = getOwn(resolvedImports, name) || getOwn(imports, name);
    if (!info && (resolvers == null ? void 0 : resolvers.length)) {
      const resolved = await firstMatchedResolver(resolvers, name);
      if (resolved) {
        if (typeof resolved === "string") {
          info = {
            path: resolved,
            importName: "default"
          };
        } else {
          info = resolved;
        }
        resolvedImports[name] = info;
      }
    }
    if (!info || !info.path)
      continue;
    addToModules({
      path: info.path,
      importName: info.importName,
      name
    });
    if (info.sideEffects) {
      const infos = [info.sideEffects].flat(1).map((info2) => {
        if (typeof info2 === "string")
          return { path: info2 };
        return info2;
      });
      infos.forEach((info2) => addToModules(info2));
    }
  }
  if (!Object.keys(modules).length)
    return;
  const importStatements = Object.entries(modules).map(([moduleName, infos]) => {
    const imports2 = [];
    const namedImports = [];
    infos.forEach(({ name, importName }) => {
      if (name) {
        if (importName === "*")
          imports2.push(`* as ${name}`);
        else
          namedImports.push(importName && name !== importName ? `${importName} as ${name}` : name);
      }
    });
    if (namedImports.length)
      imports2.push(`{ ${namedImports.join(", ")} }`);
    return `import ${imports2.length > 0 ? `${imports2.join(", ")} from ` : ""}'${moduleName}';`;
  }).join("");
  const s = new (0, _magicstring2.default)(code);
  s.prependLeft(0, importStatements);
  return {
    code: s.toString(),
    map: sourceMap ? s.generateMap({ source: id, includeContent: true }) : null
  };
}
async function firstMatchedResolver(resolvers, name) {
  for (const resolver of resolvers) {
    if (typeof resolver === "object" && resolver.type === "directive") {
      if (name.startsWith("v"))
        name = name.replace("v", "");
      else
        continue;
    }
    const resolved = await (typeof resolver === "function" ? resolver(name) : resolver.resolve(name));
    if (resolved)
      return resolved;
  }
}
var hasOwnProperty = Object.prototype.hasOwnProperty;
function getOwn(object, key) {
  return hasOwnProperty.call(object, key) ? object[key] : void 0;
}

// src/core/dts.ts
function generateDeclaration(imports, resolvedImports = {}) {
  const body = [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports)
  ].sort((a, b) => a[0].localeCompare(b[0])).map(([name, info]) => `  const ${name}: typeof import('${info.path}')${info.importName !== "*" ? `['${info.importName || name}']` : ""}`).join("\n");
  return `// Generated by 'unplugin-auto-import'
// We suggest you to commit this file into source control
declare global {
${body}
}
export {}
`;
}

// src/core/eslintrc.ts
function generateESLintConfigs(imports, resolvedImports = {}, eslintrc) {
  const eslintConfigs = { globals: {} };
  [
    ...Object.entries(imports),
    ...Object.entries(resolvedImports)
  ].sort((a, b) => a[0].localeCompare(b[0])).forEach(([name]) => {
    eslintConfigs.globals[name] = eslintrc.globalsPropValue || true;
  });
  const jsonBody = JSON.stringify(eslintConfigs, null, 2);
  return jsonBody;
}

// src/index.ts
var src_default = _unplugin.createUnplugin.call(void 0, (options) => {
  let resolved = resolveOptions(options);
  if (!Object.keys(resolved.imports).length && !resolved.resolvers.length)
    console.warn("[auto-import] plugin installed but no imports has defined, see https://github.com/antfu/unplugin-auto-import#configurations for configurations");
  const generateConfigFiles = _utils.throttle.call(void 0, 500, false, () => {
    if (resolved.dts)
      _fs.promises.writeFile(resolved.dts, generateDeclaration(resolved.imports, resolved.resolvedImports), "utf-8");
    const { eslintrc } = resolved;
    if (eslintrc.enabled && eslintrc.filepath)
      _fs.promises.writeFile(eslintrc.filepath, generateESLintConfigs(resolved.imports, resolved.resolvedImports, eslintrc), "utf-8");
  });
  return {
    name: "unplugin-auto-import",
    enforce: "post",
    transformInclude(id) {
      return resolved.idFilter(id);
    },
    async transform(code, id) {
      const res = await transform(code, id, resolved);
      if (res)
        generateConfigFiles();
      return res;
    },
    vite: {
      configResolved(config) {
        resolved = resolveOptions(options, config.root);
        generateConfigFiles();
      }
    }
  };
});



exports.src_default = src_default;
